### PR TITLE
Make typeck aware of uninhabited types

### PR DESCRIPTION
**Pull request by camsteffen**
It was previously assumed in #85556 that we can't check if an expression has an uninhabited type in typeck. So an `unreachable_code` case was added to the liveness pass instead. This PR moves that case back into typeck. It just requires adding `cycle_delay_bug` to the uninhabited query.

Incidentally this leads to some new cases for `unreachable_code`. But it should simply be consistent with how `!` is currently handled.

My original motivation is to remove a blocker to doing some refactoring on liveness.

r? rust-lang/compiler

### CODE REVIEW

**Comment by camsteffen** [comment]
I'm not entirely confident in the test changes tbh. Maybe a few corner cases need some ironing.

File: compiler/rustc_typeck/src/lib.rs
Comment (camsteffen) [reply compiler/rustc_typeck/src/lib.rs:522]: The problem is I only want to stop for a specific error.

File: compiler/rustc_typeck/src/check/check.rs
Comment (camsteffen) [reply compiler/rustc_typeck/src/check/check.rs:766]: Ah, I didn't know. Will do!

**Comment by cjgillot** [comment]
How does this PR articulate with https://github.com/rust-lang/rust/pull/100720?

**Comment by camsteffen** [comment]
There are a number of queries that are susceptible to cycle errors if they run on an infinitely recursive (unrepresentable) ADT. adt_sized_constraint and type_uninhabited_from are a couple such queries. There are multiple ways to solve the cycle error problem for a given query:

1. Ensure the query never runs on unrepresentable types. (this PR stops rustc before bodies are type-checked, so that `type_implemented_from` remains guarded from unrepresentable types)
2. Implement `Value::from_cycle_error` to have a fallback (`adt_sized_constraint` currently does this)
3. Add `if !is_representable(field_id) { return <fallback> }`. (this is an idea I had that prompted me to do #100720)

So we can either be conservative and try to guard queries from unrepresentable types as much as possible (halt rustc early), or go the other direction and try to recover-and-continue as much as possible for unrepresentable types. Or maybe the right balance is somewhere in the middle. I'm not quite sure.

In any case, it might make sense to try and merge #100720 first on its own, and then we can think about how other parts of the compiler may or may not use the new `is_representable` query. Though I think it might also be fine to merge this PR first. They are only tangentially related.

**Comment by camsteffen** [comment]
Removed the early exit and instead added `from_cycle_error` for `type_uninhabited_from`.

File: compiler/rustc_typeck/src/lib.rs
Comment (camsteffen) [reply compiler/rustc_typeck/src/lib.rs]: Removed this change.

File: compiler/rustc_query_impl/src/values.rs
Comment (cjgillot) [inline compiler/rustc_query_impl/src/values.rs:73]:
```suggestion
        // SAFETY: This is never called when `Self` is not `TypeUninhabitedFrom<'tcx>`.
```

File: src/test/ui/consts/const_discriminant.rs
Comment (oli-obk) [inline src/test/ui/consts/const_discriminant.rs:29]: this comment is now outdated.

cc @rust-lang/wg-const-eval for slight changes in const eval semantics

File: src/test/ui/break-diverging-value.rs
Comment (oli-obk) [inline src/test/ui/break-diverging-value.rs:27]: cc @rust-lang/types a small bugfix makes some code pass that didn't pass before, because it didn't realize that `get_void()`'s result means the code is unreachable.

File: src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs
Comment (oli-obk) [inline src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:287]: I think we'll want to keep the previous logic and just allow some lints.

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by oli-obk** [comment]
@bors try

let's do a crater run

File: src/test/ui/consts/const_discriminant.rs
Comment (RalfJung) [reply src/test/ui/consts/const_discriminant.rs:29]: Does this still behave as before if we instead `transmute` from `&()` to `&Void`? Right now this doesn't really test any more what it should test.

File: src/test/ui/consts/const_discriminant.rs
Comment (RalfJung) [inline src/test/ui/consts/const_discriminant.rs:1]: This makes all of `main` dead code, which is definitely not good.

File: src/test/ui/consts/const_discriminant.rs
Comment (RalfJung) [reply src/test/ui/consts/const_discriminant.rs:29]: Also if typeck properly respects visibility, the old behavior can probably be restored by moving `Void` into another module or so? That seems better (maybe also for a bunch of the other tests)

Comment (camsteffen) [reply src/test/ui/consts/const_discriminant.rs:29]: Makes sense. I guess I should add a test for when the type is visibly uninhabited then.

**Comment by oli-obk** [comment]
@craterbot check

File: src/test/ui/consts/const_discriminant.rs
Comment (RalfJung) [reply src/test/ui/consts/const_discriminant.rs:29]: Well this particular test was intended to cover the case where the type is uninhabited but the type checker doesn't see that. So it wouldn't make much sense to use visibly uninhabited types.

**Comment by oli-obk** [comment]
10 regressions where the unreachable expression lint now triggers more aggressively

48 regressions with "Failed to get rustc version" in a libc build script. No clue how that happened.

**Comment by RalfJung** [comment]
> 48 regressions with "Failed to get rustc version" in a libc build script. No clue how that happened.

That's https://github.com/rust-lang/crater/issues/663

**Comment by camsteffen** [comment]
Addressed comments in fixup commits. Is the crater run good enough?

**Comment by oli-obk** [comment]
This PR adds new UB detection to const eval, though it seems to be nothing anyone relied on (crater did not show any new const eval failures).

```rust
#[derive(Copy, Clone)]
pub struct ChildStdin {
    inner: AnonPipe,
}
#[derive(Copy, Clone)]
enum AnonPipe {}

const FOO: () = {
const FOO: () = {
    union Foo {
        a: ChildStdin,
        b: (),
    }
    let x = unsafe { Foo { b: () }.a };
    let x = &x.inner;
};
```

`FOO` now fails to be evaluated, because the first `let x` must be unreachable as `ChildStdin` cannot possibly be constructed.


There's also some code that will now pass typeck when it didn't before:

```rust
fn loop_break_void() -> i32 {
    let loop_value = loop { break get_void() };
    // used to error because typeck thought the function
    // could return and thus need to coerce `()` into `i32`
    // right here.
}
```

I propose to merge this change as the changes are what would be expected from the language side, so this is just a bugfix.

@rfcbot merge

**Comment by RalfJung** [comment]
? There's also some code that will now pass typeck when it didn't before:

Oh, this can be done without lang team involvement?

File: src/test/ui/consts/issue-64506.rs
Comment (RalfJung) [inline src/test/ui/consts/issue-64506.rs:9]: I think this means https://github.com/rust-lang/rust/issues/64506 is not tested any more. The error was in the field access which we don't reach any more.

The tests that are changed by this PR need auditing to ensure they still test what they are supposed to test.

Comment (RalfJung) [reply src/test/ui/consts/issue-64506.rs:9]: In particular I think all cases where a pass-tests becomes a fail-test are wrong, and need to be fixed. But others might be wrong, too, if they now fail in a different spot than before.

Comment (camsteffen) [reply src/test/ui/consts/issue-64506.rs]: Changed this to use a privately uninhabited type. Maybe one other concern is the 93161 test, but I think it's okay because it is still `run-pass`.

**Comment by camsteffen** [comment]
Rebase needed to address the mir-opt failures. Some new changes in those tests.

File: src/test/ui/consts/issue-64506.rs
Comment (RalfJung) [reply src/test/ui/consts/issue-64506.rs]: That one just gets some warnings, so should be fine.

File: compiler/rustc_typeck/src/check/expr.rs
Comment (camsteffen) [inline compiler/rustc_typeck/src/check/expr.rs:237]: Wondering if this should be cached on FnCtxt?

Comment (oli-obk) [reply compiler/rustc_typeck/src/check/expr.rs:237]: we can do a perf run to see if that's an issue (but I think you need to rebase first)

**Comment by camsteffen** [comment]
Try again with caching parent module and only checking uninhabited for certain ExprKinds.

@bors try @rust-timer queue

**Comment by camsteffen** [comment]
I think the uninhabited query does a lot of redundant work currently. I have some ideas to improve this, basically to look like or even depend on #10070. So this could be blocked on #10070 (again) if we want better perf.

**Comment by camsteffen** [comment]
@bors try @rust-timer queue

I added a small change to reduce uninhabited queries.

**Review by oli-obk — APPROVED** [review]

**Comment by camsteffen** [comment]
I think those perf results are okay? I also have a draft PR #102660 to further optimize uninhabited, but it may be overkill.

**Comment by oli-obk** [comment]
> I think those perf results are okay? I also have a draft PR #102660 to further optimize uninhabited, but it may be overkill.

will that, in combination with this PR reduce the number of query entries in the cache? Because from the perf regression it looks like that's the cause of the regression. Maybe try running your other PR on top of this one?

I'd like to at least know what additional query invocations are happening now and why they get invalidated or recomputed a lot.

**Comment by camsteffen** [comment]
> will that, in combination with this PR reduce the number of query entries in the cache?

I'm hoping so. The main idea with the uninhabited rewrite is to remove ParamEnv from the query key. Maybe running the query at different stages leads to a lot of different ParamEnvs?

> Maybe try running your other PR on top of this one?

Sure, let's give it a shot (after CI is green).

File: compiler/rustc_middle/src/ty/inhabitedness/mod.rs
Comment (camsteffen) [inline compiler/rustc_middle/src/ty/inhabitedness/mod.rs:118]: please advise :)

Comment (camsteffen) [reply compiler/rustc_middle/src/ty/inhabitedness/mod.rs:118]: That is, can I use `ParamEnv::empty()` to detect simple cases? (like when array length is a literal)

**Comment by camsteffen** [comment]
Yeehaw that's pretty good! Just need to update some code comments.

File: compiler/rustc_middle/src/ty/inhabitedness/mod.rs
Comment (oli-obk) [reply compiler/rustc_middle/src/ty/inhabitedness/mod.rs:118]: I think what you want for that case is `len.kind().ty_to_machine_usize(tcx)`. This will not attempt evaluation, but just grab an already evaluated value

Comment (camsteffen) [reply compiler/rustc_middle/src/ty/inhabitedness/mod.rs:118]: Yes perfect.

**Comment by camsteffen** [comment]
Blocked on https://github.com/rust-lang/rust/pull/102660

**Comment by camsteffen** [comment]
Bumped into #103220. I may need help with this one.

**Comment by cjgillot** [comment]
As lifetimes are not meaningful in this check, you can call `tcx.erase_regions` on the types before calling the query.

**Comment by camsteffen** [comment]
> As lifetimes are not meaningful in this check, you can call `tcx.erase_regions` on the types before calling the query.

Then I bump into `type variables should not be hashed`. Is `freshen` (last commit) good here?

**Comment by compiler-errors** [comment]
I'm not too familiar with this PR so take my advice with a grain of salt, but if you're performing a query on a type that has inference variables, and then returning a response that relates to those inference variables, you really should be canonicalizing: https://doc.rust-lang.org/stable/nightly-rustc/rustc_infer/infer/canonical/canonicalizer/struct.Canonicalizer.html#method.canonicalize

**Comment by cjgillot** [comment]
I think that ICE saved us from an hard-to-debug issue.
The current impl of `inhabited_predicate_type` does not handle type variables.  So having the query system preventing us from doing that is a good thing.

There are 2 cases:
- either we are sufficiently late in the type-checking and we can resolve the variables --> call `tcx.erase_regions(self.resolve_vars_if_possible(ty))`;
- either we are not, the above call will still contain inference variables, and the check will have to be skipped.

IIUC, the goal of the call to `inhabited_predicate_type` is to diagnose unreachable expressions. Can this check be delayed after type inference? Delaying this will allow to use `self.fully_resolve` which always returns a fully resolved type, suitable for the query.

**Comment by camsteffen** [comment]
> IIUC, the goal of the call to `inhabited_predicate_type` is to diagnose unreachable expressions. Can this check be delayed after type inference?

If it's just for the `unreachable_code` lint then that would be easy enough. But this affects typeck a little bit more than that. See [here](https://github.com/rust-lang/rust/blob/758f19645b8ebce61ea52d1f6672fd057bc8dbee/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs#L1449) and the change to `src/test/ui/break-diverging-value.rs`.

**Comment by cjgillot** [comment]
After thinking a bit more about what happens, I agree that using `freshen` is the right thing to do. It will infer variables where possible, erase lifetimes, and replace remaining infer vars by fresh vars that are ok to be hashed.
The analysis is a bit more conservative than it could as it will consider fresh infer types as inhabited, which is probably not an issue.
The failure looks unrelated to the typeck changes, probably just needs re-blessing a `ui/deriving` test.

Splitting the `unreachable code` lint out may be a good idea, but that's out of the scope of this PR.

**Comment by camsteffen** [comment]
> The analysis is a bit more conservative than it could as it will consider fresh infer types as inhabited, which is probably not an issue.

Is this really okay? I'm worried that such cases are not well defined. Like if we make `resolve_vars_if_possible` "smarter" in the future, we will inadvertently change some behavior with uninhabited types.

> Splitting the `unreachable code` lint out may be a good idea, but that's out of the scope of this PR.

Agreed.

**Comment by cjgillot** [comment]
resolve_vars_if_possible probably won't get smarter, but it's results depend on the unification state of the type checker. Notably, on the order in which expressions are checked for type inference. I'll have to read the surrounding code more precisely to see how that order interferes with the uninhabited check.

**Comment by camsteffen** [comment]
Unfortunately I haven't had much time for rust lately. @cjgillot PM'ed me and offered to take over here. I'm just going to close this PR.

**Pull request closed without merging**
